### PR TITLE
feat: add null-as-missing gap filling

### LINKED ISSUE #7177: `FILL(PREVIOUS)` gap filling (treat_null_as_missing) in `GapFillExec`
Part of gap-filling implementation: #6594.

Requires #6973 to be implemented first.

This issue adds the ability to fill gaps with the previous value in aggregate column if there was any, where null values in the input column are treated as missing and are filled with the previous non-null value, if any.

This implements gap-filling behavior equivalent to InfluxQL's `FILL(PREVIOUS)`.

This documentation for `LOCF` is useful for context:
http://docs-v3-dev.timescale.com.s3-website-us-east-1.amazonaws.com/docs-replication-landingpage-lana/api/replication-landingpage-lana/analytics/locf/

The work of this issue is to implement `LOCF(treat_null_as_missing=true)` semantics in `GapFillExec`.

### PR DESCRIPTION
**Pull request by wolffcm**
Closes #7177.

This adds a new mode of gap filling to `GapFillExec`, which will fill both gaps and `null` values in input columns with previous non-null values. This is equivalent to `fill(previous)` in InfluxQL.

I also made several changes to the tests in `algo.rs`:
- I removed multi-output batch tests, as they are painful to maintain. I found them useful when prototyping but now they are redundant with tests in `exec_tests.rs`, which exercise `GapFillExec` with multiple input and output batch sizes.
- I updated some existing tests to use `insta` snapshots and generally be easier to read.

### CODE REVIEW

File: iox_query/src/exec/gapfill/algo.rs [hunk 1]
```diff
@@ -225,33 +225,16 @@ impl GapFiller {
...
-                Some(fs) => Err(DataFusionError::NotImplemented(format!(
-                    "unsupported gap fill strategy {fs:?}"
-                ))),
-                None => Err(DataFusionError::Internal(format!(
-                    "could not find fill strategy for aggregate column with index {idx}"
-                ))),
-            }?;
-            if take_vec.len() != output_time_len {
+            let output_array =
+                cursor.build_aggr_col(&self.params, series_ends, input_time_array, aa)?;
```
Comment (wolffcm) [inline iox_query/src/exec/gapfill/algo.rs:229]: I moved this logic into `Cursor` itself, since now output arrays for aggregate columns may be produced using either the `take` or `interleave` kernel.

File: iox_query/src/exec/gapfill/algo.rs [hunk 2]
```diff
@@ -512,15 +496,45 @@ impl Cursor {
...
+                input_aggr_array,
+            ),
+        }
+    }
+
+    /// Builds an array using the [`take`](take::take) kernel
     /// to produce an aggregate output column, filling gaps with
     /// null values.
-    fn build_aggr_take_vec_fill_null(
+    fn build_aggr_fill_null(
```
Comment (wolffcm) [inline iox_query/src/exec/gapfill/algo.rs:531]: I also refactored these methods to actually build their output arrays using the appropriate kernel, since either `take` or `interleave` may be used.

File: iox_query/src/exec/gapfill/algo.rs [hunk 3]
```diff
@@ -917,8 +1117,9 @@ mod tests {
     }
 
     #[test]
-    fn test_cursor_append_aggr_take() -> Result<()> {
```
Comment (wolffcm) [inline iox_query/src/exec/gapfill/algo.rs:920]: All of these tests are unchanged functionally, I just had to update them to reflect how `Cursor` was refactored, and I also took the opportunity to use `insta` snapshots as well.

File: iox_query/src/exec/gapfill/algo.rs [hunk 4]
```diff
@@ -1096,424 +1343,203 @@ mod tests {
...
     #[test]
-    fn test_cursor_multi_output_batch() -> Result<()> {
-        let output_batch_size = 5;
-        let params = GapFillParams {
-            stride: 50,
-            first_ts: Some(950),
-            last_ts: 1350,
-            fill_strategy: simple_fill_strategy(),
-        };
+    fn test_cursor_aggr_prev_null_as_missing() {
```
Comment (wolffcm) [inline iox_query/src/exec/gapfill/algo.rs:1375]: The diff is not very pretty, but this the only new test here.

File: iox_query/src/exec/gapfill/algo.rs [hunk 5]
```diff
@@ -579,20 +601,73 @@ impl Cursor {
...
+    /// previous values in the column.
+    fn build_aggr_fill_prev_stashed(
+        &mut self,
+        params: &GapFillParams,
+        series_ends: &[usize],
+        input_time_array: &TimestampNanosecondArray,
+        input_aggr_array: &ArrayRef,
+    ) -> Result<ArrayRef> {
+        let stash = self.get_aggr_col_state().stash();
+        let mut aggr_builder = StashedAggrBuilder {
```
Comment (alamb) [inline iox_query/src/exec/gapfill/algo.rs:646]: using interleave is nice.

File: iox_query/src/exec/gapfill/algo.rs [hunk 6]
```diff
@@ -787,13 +896,104 @@ enum RowStatus {
...
+impl StashedAggrBuilder<'_> {
+    /// Create a 2-element array containing a null value and the value from
+    /// `input_aggr_array` at `offset` for use with the [`interleave`](arrow::compute::interleave)
+    /// kernel.
+    fn create_stash(input_aggr_array: &ArrayRef, offset: u64) -> Result<ArrayRef> {
+        let take_arr = vec![None, Some(offset)].into();
+        let stash =
+            take::take(input_aggr_array, &take_arr, None).map_err(DataFusionError::ArrowError)?;
+        Ok(stash)
+    }
```
Comment (alamb) [inline iox_query/src/exec/gapfill/algo.rs:919]: I wonder if we could avoid this copy into a new intermediate array -- couldn't we keep the old array reference + offset around (as we are using the `interleave` kernel anyways)

File: iox_query/src/exec/gapfill/algo.rs [hunk 7]
```diff
@@ -1096,424 +1343,203 @@ mod tests {
...
-        // with the last null being the first element of the second batch.
-        test_helpers::maybe_start_logging();
-        let output_batch_size = 4;
-        let params = GapFillParams {
-            stride: 50,
-            first_ts: Some(1000),
-            last_ts: 1100,
-            fill_strategy: simple_fill_strategy(),
-        };
+    fn test_cursor_aggr_prev_null_as_missing_stashed() {
```
Comment (alamb) [inline iox_query/src/exec/gapfill/algo.rs:1439]: Rather than getting this exact state, I wonder if you could extend the other tests to be "fuzz" like and ensure that the output is the same when the inputs are sliced

in psuedo code:
1. for `offset` in `0..input_agr_arr.len()` 
2.   let input_aggr_array1 = input_aggr_arr.slice(0, offset);
3.  let input_aggr_array2 = input_aggr_arr.slice(offset, len);
4. Gap fill with `input_aggr_array1` and `input_aggr_array2`

I am not sure this is the right level to do that sort of slicing test as it could get quite complicated 🤔

**Review by alamb — APPROVED** [review]
I found this PR very easy to read and review -- the structure makes sense, the comments are great, as are the tests. Thank you very much.

As an aside, the `interleave` kernel makes me think that perhaps we could use that approach in the `Deduplicate` kernel to improve performance (rather than `concat_batches`) -- maybe this is something @tustvold  has been trying to tell us 🤔

File: iox_query/src/exec/gapfill/algo.rs (hunk 6 above)
Comment (wolffcm) [reply iox_query/src/exec/gapfill/algo.rs:919]: It might be possible. A couple points to note though:
- We do need a `null` element, since unlike `take`, `interleave` lets you only create an output array using elements from the input arrays. There should be one handy though at `offset + 1`, since we would not need to stash otherwise.
- We could keep the old array around, but it would make it harder to track the memory, since the memory tracking is done in `GapFillStream`, not here. (I am not tracking the memory allocated here since it's a single element---maybe I should though)
- I've made an effort to only buffer input rows that are needed to produce an output batch, and have a bound that is roughly `output_batch_size + 2` (but of course this depends on the size of the batches we get from the input stream, so it will be more). Avoiding a copy here would increase that footprint, not sure if it would be worth it.

File: iox_query/src/exec/gapfill/algo.rs (hunk 7 above)
Comment (wolffcm) [reply iox_query/src/exec/gapfill/algo.rs:1439]: I think the tests in `exec_tests.rs` cover the cases you're describing. By setting the `output_batch_size` to 1, all of the edge cases and possible states will be covered, given the right input.

I've been using these test during intermediate stages until I'm ready to write a test against the `ExecutionPlan` interface, at which point `exec_tests.rs` is generally more useful.

File: iox_query/src/exec/gapfill/exec_tests.rs [hunk 8]
```diff
@@ -573,6 +573,185 @@ fn test_gapfill_fill_prev() {
...
+            assert_batch_count(&batches, output_batch_size);
+        }
+    }}
+}
+
+#[test]
+fn test_gapfill_fill_prev_null_as_missing_many_nulls() {
+    test_helpers::maybe_start_logging();
+    insta::allow_duplicates! {
+        for output_batch_size in [16, 1] {
```
Comment (wolffcm) [inline iox_query/src/exec/gapfill/exec_tests.rs:658]: @alamb By setting `output_batch_size` here I think this achieves the "fuzz" style testing that you mentioned above.

**Comment by wolffcm** [comment]
> As an aside, the interleave kernel makes me think that perhaps we could use that approach in the Deduplicate kernel to improve performance (rather than concat_batches) -- maybe this is something @tustvold has been trying to tell us 🤔

It never occurred to me to use `interleave` as a way to avoid having to use `concat`, which seems like it must be costly. I really like this idea! It adds some additional complexity, but it seems like this operator could really benefit from using `interleave` over `take` more generally.

File: iox_query/src/exec/gapfill/algo.rs (hunk 6 above)
Comment (alamb) [reply iox_query/src/exec/gapfill/algo.rs:919]: I agree it is not worth pursuing at this time -- thank you for considering it.

File: iox_query/src/exec/gapfill/algo.rs (hunk 7 above)
Comment (alamb) [reply iox_query/src/exec/gapfill/algo.rs:1439]: Makes sense

**Comment by alamb** [comment]
> as a way to avoid having to use concat, which seems like it must be costly. I really like this idea!

I think in particular `concat` may result in dictionaries with many repeated values (aka the dictionary becomes huge). @tustvold  and I have discussed this as a theoretical concern for a while but I don't think we have any "smoking gun" evidence / measurement in our actual customer workloads

Here is one attempt to reduce improve the situation (re-encode dictionaries): https://github.com/apache/arrow-rs/pull/3558

**Comment by tustvold** [comment]
TBC iterleave currently behaves the same way as concat with respect to dictionaries, but if you can use it to avoid chaining its output into the next input, that would yield benefits

**Comment by wolffcm** [comment]
> Maybe now that it uses interleave as well, it should optimize afterwards?

Actually, never mind that part, since it's interleaving with just a two-element array, the difference in the dictionary columns would be trivial.
